Box advection test

trontrytel opened the pull request:
# Description

This is a box advection test that I wrote using SingleStack configuration to visualise different filter options. I wrote it so that I could learn how different filters work. Maybe it could be useful for testing oscillation limiters?

It can generate plots like this:

![Filter_options jpeg 001](https://user-images.githubusercontent.com/1196696/88236628-4a960c00-cc32-11ea-8f87-52e77e028725.jpeg)

If it's useful I can turn it into a proper test or tutorial. Let me know what you think.

<!--- Please fill out the following section --->

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

<!--- Please leave the following section --->

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

thomasgibson commented:
I think it's a good idea. Can you add the Boyd-Vandeven filter to this set of experiments?

thomasgibson commented on tutorials/Numerics/DGMethods/box1D.jl:
```diff
@@ -0,0 +1,301 @@
+                ExponentialFilter(solver_config.dg.grid, exp_param_1, exp_param_2)
+            )
+            nothing
+        end
```
Can you add the Boyd-Vandeven filter to this too?

trontrytel replied:
@thomasgibson - I'm adding it now. I guess it should also be added to the filter tests in `ClimateMachine.jl/test/Numerics/Mesh/filter.jl`. But I don't know how to test it.

tapios commented:
Super useful! It'll be nice to use this to test the Roe fluxes with it when we have them together. They should provide a substantial improvement on the oscillations.

lcw commented:
Looks good to me!

My only suggestion is that you may want to add an option to the plotting (if it is not there already) to sample the solution on a fine grid.  Like what was done in https://github.com/CliMA/ClimateMachine.jl/commit/4514acad6b696ef240c85a888599991bbacd5b71 for VTK.  Then, my guess is you would still see the wiggles around zero in the TMAR filter cases.

trontrytel commented:
> Looks good to me!
> 
> My only suggestion is that you may want to add an option to the plotting (if it is not there already) to sample the solution on a fine grid. Like what was done in [4514aca](https://github.com/CliMA/ClimateMachine.jl/commit/4514acad6b696ef240c85a888599991bbacd5b71) for VTK. Then, my guess is you would still see the wiggles around zero in the TMAR filter cases.

Thanks @lcw! 

I'm piggybacking here on whatever is done in `SingleStackUtils.jl` @charleskawczynski - Could we add the same sampling option to the output there? 

@lcw - Way back when you wrote (and I watched) the test for the TMAR filter that somehow ended up in tutorials: https://github.com/CliMA/ClimateMachine.jl/tree/master/tutorials/Numerics/DGMethods/nonnegative.jl. Could I move it to tests and collect some plots from the box setup as tutorial?

lcw commented:
Anna Jaruga <notifications@github.com> writes:

> I'm piggybacking here on whatever is done in `SingleStackUtils.jl`
> @charleskawczynski - Could we add the same sampling option to the
> output there?

Oh cool, I haven't used `SingleStackUtils.jl`.

> @lcw - Way back when you wrote (and I watched) the test for the TMAR
> filter that somehow ended up in tutorials:
> https://github.com/CliMA/ClimateMachine.jl/tree/master/tutorials/Numerics/DGMethods/nonnegative.jl.
> Could I move it to tests and collect some plots from the box setup as
> tutorial?

Sure that sounds great.

vchuravy commented on tutorials/Numerics/DGMethods/showcase_filters.jl:
```diff
@@ -3,7 +3,7 @@
 # See [Filters API](https://clima.github.io/ClimateMachine.jl/latest/APIs/Numerics/Meshes/Mesh/#Filters-1) for filters interface details.
 
-include("Box1D.jl")
+include(joinpath("tutorials", "Numerics", "DGMEthods", "Box1D.jl"))
```
`@__DIR__` to get the current path of the file.

trontrytel commented on tutorials/Numerics/DGMethods/Box1D.jl:
```diff
@@ -0,0 +1,305 @@
+    init_state_prognostic!,
+    boundary_state!
+
+ClimateMachine.init(; disable_gpu = true, log_level = "warn");
```
@charleskawczynski - look here

charleskawczynski commented:
I think including a filtering/sampling would be good for a separate PR, and perhaps better suited for LES config (I believe our hope is to have zero horizontal variations for single stack configurations).

trontrytel commented:
@charleskawczynski or anyone else powerful enough to do `r+`

I squashed, rebased and opened an issue for the sampling option. Unfortunately the `@Info` clutter does not disappear in the tutorial page in the same way as it did in my terminal output locally. 

Otherwise I think it's good to be merged. Thank you for all the feedback!

charleskawczynski commented:
bors r+

charleskawczynski commented:
bors r+

trontrytel commented:
bors r+
